### ISSUE TITLE
Source context not being correctly configured during MVM Deployment

### ISSUE THREAD

**Issue by patflynn** [GoogleCloudPlatform/cloud-code-intellij#450]
Apparently from a convo with Erez this should work now.

What I've found is you need to call the gcloud command to generate the src context directly, and I still was not able to get CDB to work.

More investigation is required to package the src ctxt correctly for a custom mvm deployment.

If we do a maven or gradle build as part of the deployment (in the 'run before' model), ideally this is when we'd generate the src context, so I can imagine it being a 'run before' task that we auto configure.

random other issue:  If deploy does eventually automatically just work, we still would not work because it it requires the deployment staging directory to be in the git project. Currently we create a staging directory in the user's tmp directory.

**Comment by patflynn** [comment]
IMO, tackle this last as the experience works without it.

**Comment by chanseokoh** [comment]
Does generating the src context files have to be associated with the deployment process? One could think of this outside the context of deployment. For example, our cloud plugin could always generate and add the src context files whenever building a WAR inside the IDE.

If we are going to associate this source-context-generation with deployment, then I think I would go with generating and adding the src context files to an already packaged WAR that is ready to be deployed, i.e., unpacking the WAR, generating the files, and re-packaging them in the staging area.

**Comment by patflynn** [comment]
does the context need to be packaged into the war for it to work?

**Comment by etanshaul** [comment]
https://cloud.google.com/debugger/docs/setting-up-java-on-app-engine explains that the source context file should go in WEB-INF so another question to answer is if there is a way for this to work with jar files.

**Comment by patflynn** [comment]
That's the appcfg.sh flow.  I'm not sure that gcloud does the same.

**Comment by chanseokoh** [comment]
I think it is how it works. For example, you would manually generate them anyway and put them under WEB-INF.

https://cloud.google.com/debugger/docs/setting-up-on-compute-engine#build_test_and_deploy_your_application

**Comment by chanseokoh** [comment]
Okay, we will need to call 'gcloud preview app gen-repo-info-file' to generate the src context files at some point.

Initially, I was thinking I could do that just before calling 'gcloud preview app deploy', as part of our IntelliJ deployment process. @patflynn, you seem to have something else in mind. Can you elaborate on what you said last week, about the 'run before' (IDE/gradle/maven?) task?

Next, the only requirement is that those context files be in class path. One option is to inject those context files into a WAR (WEB-INF/classes) or a JAR (root directory) directly. Another option would be to place them somewhere in a VM and include the directly with CLASSPATH?

**Comment by etanshaul** [comment]
moving off of 0.9.3 release because CDB does not currently work in MVM.

**Comment by chanseokoh** [comment]
This is on hold because, unlike what it's supposed to do, 'gcloud deploy' currently doesn't automatically pick up source context files if they are in the same directory where a packaged WAR/JAR exists (internal bug number: 28317242). We also don't want to inject the context files to an existing WAR/JAR ourselves.

**Comment by etanshaul** [comment]
@chanseokoh do you know what the status of this is? Are we blocked on this still?

**Comment by chanseokoh** [comment]
Yes, it depends on the bug in the above https://github.com/GoogleCloudPlatform/gcloud-intellij/issues/450#issuecomment-214457248.

**Comment by etanshaul** [comment]
Ok thanks.

**Comment by joaoandremartins** [comment]
@patflynn @chanseokoh was there ever a decision on how gen-repo-info-file should be invoked? i.e., with a Before Launch task or implicitly in a deployment? Judging from @chanseokoh's email, apparently not...

I've got a strong preference for doing this implicitly, to reduce UI complexity. It's not immediately clear to the user what source-context is, or what purpose it serves. There is also the chance that a user just deletes it from the Before Launch task list because they don't understand what it's for.

**Comment by chanseokoh** [comment]
@joaoandremartins not answering your question but just throwing some more info:

For Java on GCE, there are Debugger setup instructions where it is documented how to generate the source context files:

https://cloud.google.com/debugger/docs/setting-up-on-compute-engine#generate_the_source_context_file

For GCE, users do have to manually generate the files in order to use the Debugger. The only requirement then would be to to put the files in places where a Java process can find them or simply in the classpath for that purpose. (So, there might be advanced users who couple these files with their WARs, putting under `WEB-INF/classes`, as part of their own release process.)

`appcfg.sh` does all this magically (i.e., it generates and packages the files as part of deploy): https://cloud.google.com/debugger/docs/setting-up-java-on-app-engine. There was a discussion whether `gcloud app deploy` should do the same thing, but we reached the consensus that it won't. (Maybe the reason was to be language-agonistic?)

**Comment by joaoandremartins** [comment]
Ah thanks for the pointer, @chanseokoh. I guess if we documented source context has to be explicitly generated (for GCE/Flex), then it makes sense to have it in the Before Launch tasks. A user who removes it from the list either doesn't need it, or knows what they're doing.

**Comment by chanseokoh** [comment]
> For GCE, users do have to manually generate the files in order to ~~use the Debugger~~ for UI to be able to point to the corresponding source.

Just corrected what I said above. To clarify, Debugger works regardless of whether the source context files exist or not. It's just for making any UI to know which source it should point to.

**Comment by aslo** [comment]
Just to clarify, are you saying that the steps from part 1 will occur automatically by default? Or will the user have to set those up?

**Comment by joaoandremartins** [comment]
Automatically, by default. However, the task will be visible to the user and they will be able to modify it. When they do, they're going to see something like that.

**Comment by etanshaul** [comment]
Just wondering, what's the advantage of having this exposed? Couldn't this be done automatically behind the scenes?

**Comment by joaoandremartins** [comment]
Regarding the location source context files are generated, they can either be generated to the source and then carried over by staging (staging doesn't move over source context files currently), or directly to the staged environment. Is there any recommended location we should be generating them at?

**Comment by patflynn** [comment]
do you mean where you should place the src context? I wouldn't put it in
the src tree since you don't want the user to see it there. It should be
run in the context of the git workspace with the output placed in the build
path and then packaged into the staging directory.

**Comment by joaoandremartins** [comment]
@etanshaul since we're adding files to a user's app, it kinda makes sense to me to show that to the user. If they want to disable it, they can. I thought gen-repo-info-file fits the Before Launch task model really well.

**Comment by etanshaul** [comment]
Yeah makes sense to me. 

Another option - I'm not sure yet which is better - is to use the facet to control whether or not source context is generated. Then just do the generation behind the scenes if so. For flex this would first require us to create a facet. 

I'm leaning towards your approach though.

**Comment by etanshaul** [comment]
Hopefully you can customize the text that shows up in the before launch task to make it clearer what its doing.

**Comment by joaoandremartins** [comment]
I can change the "gcloud" bit of text to something like "gcloud app gen-repo-info-file"...

**Comment by joaoandremartins** [comment]
Just to scope the work here, what I see is needed is:
- Add gen-repo-info-file action to appengine-plugins-core, which generates source context in a location (to be decided),
- Decide on the source context output location. WAR/JAR root?
- Call gen-repo-info-file on Flex staging, or as part of Flex deployment.

Things seem to be working for standard. Anything I'm missing?

Also, to wrap on the previous discussion, this will be done silently and no before launch task will be added.

**Comment by etanshaul** [comment]
That sounds right to me.

Just wondering - what is gen-repo-info-file? I thought it was gcloud app gen-config?

**Comment by joaoandremartins** [comment]
gen-config seems to be deprecated, and I wasn't sure if it also created source contexts file.
I think the recommended command to generate source context files is gen-repo-info-file.

**Comment by chanseokoh** [comment]
gen-config creates app.yaml , Dockerfile, etc. I also heard that there exists something like "gcloud beta debug gen-repo-info-file" which could probably be just "gcloud app gen-repo-info-file".

**Comment by patflynn** [comment]
Yeah gen-config is not relevant. Although I also am not aware of it being
deprecated as it's a key component of the flex deployment flow.

**Comment by patflynn** [comment]
+danielsb

I think that the library function should basically function as a pass
through that allows the specification of:
- output-directory
- source-directory

It should provide the signals that:
- the source directory is not a git workspace
- the source directory contains un-committed changes

Currently the gcloud command does error if run outside of a git workspace,
but I don't think it currently has a way of reporting whether there are
uncommitted changes.

**Comment by b-daniels** [comment]
We moved gen-repo-info-file under the debug source group because it's applicable to non-App Engine platforms. It's a good idea for gen-repo-info-file to generate a warning on uncommitted changes. I'll create a feature request and try to get it into a near future release.

**Comment by joaoandremartins** [comment]
An issue of adding source context to the artifact root is that I don't know how to "hook into" IJs package/build steps.
IJ seems to allow static files to be added and there is a "Pre-processing" section, but it only allows Ant targets to be ran.

[I asked a question on IJs forum](https://intellij-support.jetbrains.com/hc/en-us/community/posts/203374204-Adding-dynamically-generated-file-to-artifact-root). In the meantime, what other options for Flex do we have? Where is source context being picked up from now? @b-daniels

**Comment by b-daniels** [comment]
Currently we recommend WEB-INF/classes, but anything in the classpath
should work. So you can run "gcloud beta debug source gen-repo-info-file
--output-directory=<anywhere on the classpath>", or you can build the
source context files anywhere and inject them into the jar.

My current top priority is to improve that situation so that we can also
get the contexts from the jar directory. I don't have an ETA on
availability for that yet.

**Comment by joaoandremartins** [comment]
Getting source context from the jar directory is my favourite one since, apparently, to have it appear inside the jar/war, it needs to be in the source, which we didn't want to.
I can't find how to hook into the IJ/Maven packaging phase to add source context generation there. Packaging seems to be an atomic step, and we can do things before/after it, but not while (apart from those IJ Artifact output/pre-processing menus, that don't work with non-static files).
